File: docs-web.resources/com.ibm.docs.web.resources/WebContent/js/concord/chart/model/DataSeries.js
/* ***************************************************************** */
/*                                                                   */
/* IBM Confidential                                                  */
/*                                                                   */
/* IBM Docs Source Materials                                         */
/*                                                                   */
/* (c) Copyright IBM Corporation 2012. All Rights Reserved.          */
/*                                                                   */
/* ***************************************************************** */

dojo.provide("concord.chart.model.DataSeries");
dojo.require("concord.chart.utils.Utils");
dojo.declare("concord.chart.model.DataSeries", null,{
	
	/**
	 * data is a map, its keyset maybe: category, label, x, y, bubbleSize, stock?
	 * the value is DataSequence
	 */
	data: null,
	parent: null,
	_chart: null,
	
	graphPro: null,
	dataPoints: null,
	seriesId: null,
	marker: null,
	smooth: null,
	hide: false,
	/*
	 * a number, each bit represent a changed item
	 * 0: 0x01, value change
	 * 1: 0x02, series name change 
	 */
	changes: 0,
	utils : concord.chart.utils.Utils,
		
	_roles: ["label","xVal","yVal", "val", "bubbleSize"],
	
	constructor: function(plot)
	{
		if(plot!=null)
		{
			this.setParent(plot);
		}
		this.data = {};
	},
	
	setParent: function(plot)
	{
		this.parent = plot;
		this._chart = plot._chart;
	},
	
	getType: function()
	{
		return this.parent.type;
	},
	
	putDataSequence: function(key, seq)
	{
		this.data[key] = seq;
		seq.setParent(this);
		seq.setProperty("role",key);
	},
	
	getDataSequence: function(key)
	{
		return this.data[key];
	},
	
	getRoleInfo: function(role,bMSFmt,separator, bDlg)
	{
		var seq = this.data[role];
		if(seq==null)
			return null;
		var addr = seq.getAddress(bMSFmt,separator, bDlg);
		if(addr){
			if(addr == "#MULT")
				return "";
			return addr;
		}
		return seq.getData();
	},
	
	isNameDirty: function()
	{
		if(this.changes & 0x02)
			return true;
		var seq = this.data["label"];
		if(seq!=null)
			return seq.isDirty();
		return false;
	},
	
	getName: function()
	{
		var seq = this.data["label"];
		if(seq!=null)
		{
			var data = seq.getData();
			if(dojo.isArray(data)){
				var arr = data.length > 50 ? data.slice(0,50) : data;
				data = arr.join(" ");
			}
			return data;
		}
		return null;
	},
	
	isDataDirty: function()
	{
		if(this.changes & 0x01)
			return true;
		for(var role in this.data)
		{
			if(role=="label" || role=="cat")
				continue;
			var seq = this.data[role];
			if(seq.isDirty())
				return true;
		}
		return false;
	},
	
	toJson: function(noData)
	{
		var json = {};
		json["id"] = this.seriesId;
		for(var role in this.data)
		{
			var seq = this.data[role];
			if(seq!=null)
			{
				var rojson = {};
				
				var ref = seq.getAddress();
				if(ref && ref.length > 0)
					rojson.ref = ref;
				
				var data =  seq.getData();
				if(data!=null)
				{
					if(role != "label")
					{
						rojson.cache = {};
						rojson.cache.ptCount = data.length;
						if(!noData || !rojson.ref){
							if(role == "yVal" || role == "xVal")
							{
								//change null to 0 because excel doesn't support null in independent data
								rojson.cache.pts = [];
								for(var i = 0; i < rojson.cache.ptCount; i++)
								{
									if(data[i] == null)
										rojson.cache.pts.push(0);
									else
										rojson.cache.pts.push(data[i]);
								}
							}
							else
								rojson.cache.pts = data;
						}
					}else if(!noData || !rojson.ref){
						var v = data.join(" ");
						if(v.length > 0)
							rojson.v = v;
					}
				}
				
				var exRef = seq.getExRef();
				if(exRef && exRef.length > 0)
					rojson.exRef = exRef;
				json[role] = rojson;
			}
		}
		
		if(this.dataPoints!=null)
			json["dPt"] = dojo.clone(this.dataPoints);
		if(this.graphPro!=null)
			json["spPr"] = dojo.clone(this.graphPro);
		if(this.marker != null)
			json["marker"] = this.marker;
		if(this.smooth !=null)
			json["smooth"] = this.smooth;
		return json;
	},
	
	loadFromJson: function(content,dataProvider)
	{
		this.seriesId = content["id"];
		
		for(var i=0;i<this._roles.length;i++)
		{
			var role = this._roles[i];
			var json = content[role];
			if(json != null){
				var pts = null;
				if(json.cache)
					pts = json.cache.pts;
				var exRef = json.exRef;
				
				if(role == "label" && "v" in json){
					pts = [];
					pts.push(json.v);
				}
				var seq = dataProvider.createDataSequence({ref: json.ref, "role" : role, "pts" : pts, "exRef" : exRef});
				this.putDataSequence(role,seq);
			}
		}
		
		this.dataPoints = content["dPt"];
		this.graphPro = content["spPr"];
		this.marker = content.marker;
		this.smooth = content.smooth;
	},
	
	set: function(pro)
	{
		if(pro==null)
			return;
		
		var dataProvider = this._chart._dataProvider;
		for(var role in pro.data)
		{
			this._chart.controller.checkData = true;
			var json = pro.data[role];
			if(role=="label")
				this.changes |= 0x02;
			else
				this.changes |= 0x01;
			var old = this.data[role];
			if(old!=null)
				old.destroy();
			if(!json)
				delete this.data[role];
			else
			{
				var pts = null;
				if(json.cache)
					pts = json.cache.pts;
				
				if(role == "label" && "v" in json)
				{
					pts = [];
					pts.push(json.v);
				}
				var seq = dataProvider.createDataSequence({ref:json.ref, "role" : role, "pts" : pts});
				this.putDataSequence(role,seq);
			}
		}
		
		if("spPr" in pro)
		{
			if(pro.spPr==null)
				this.graphPro = null;
			else
			{
				if(this.graphPro==null)
					this.graphPro = pro.spPr;
				else
					this.utils.mergeSpPr(this.graphPro,pro.spPr);
			}
		}
		
		if("marker" in pro)
			this.marker = pro.marker;
		
		if("smooth" in pro)
			this.smooth = pro.smooth;
		
		if("dPt" in pro)
		{
			if(this.dataPoints==null)
				this.dataPoints = pro.dPt;
			else
			{
				var dpt = pro.dPt;
				if(dpt==null)
					this.dataPoints = null;
				else
				{
					for(var idx in dpt)
					{
						var dp = dpt[idx];
						if(dp==null)
							delete this.dataPoints[idx];
						else
						{
							var old_dp = this.dataPoints[idx];
							if(old_dp==null)
								this.dataPoints[idx] = dp;
							else
							{
								var old_spPr = old_dp.spPr;
								var spPr = dp.spPr;
								this.utils.mergeSpPr(old_spPr,spPr);
							}
						}
					}
				}
			}
		}
	},
	
	getReverseSettings: function(settings)
	{
		var undoPro = {};
		var data = settings.data;
		if(data!=null)
			undoPro.data = {};
		for(var role in data)
		{
			var seq = this.getDataSequence(role);
			if(seq==null)
			{
				undoPro.data[role] = "";
			}
			else
			{
				var address = seq.getAddress();
				if(address)
					undoPro.data[role] = {ref:address};
				else
				{
					var data = seq.getData();
					if(role=="label")
						undoPro.data[role] = {v:data.join(" ")};
					else
						undoPro.data[role] = {cache:{pts:data}};
				}
			}
		}
		
		if("marker" in settings)
			undoPro.marker = this.marker;
		
		if("smooth" in settings)
			undoPro.smooth = this.smooth;
		
		if("spPr" in settings)
		{
			if(this.graphPro==null)
				undoPro.spPr = null;
			else
				undoPro.spPr = this.utils.reverseSpPr(settings.spPr,this.graphPro);
		}
		if("dPt" in settings)
		{
			if(this.dataPoints==null)
				undoPro.dPt = null;
			else
			{
				var dpt = settings.dPt;
				if(dpt==null)
					undoPro.dPt = dojo.clone(this.dataPoints);
				else
				{
					undoPro.dPt = {};
					for(var idx in dpt)
					{
						var old_dp = this.dataPoints[idx];
						if(old_dp==null)
							undoPro.dPt[idx] = null;
						else
						{
							if(dpt[idx]==null || dpt[idx].spPr==null)
								undoPro.dPt[idx] = old_dp;
							else
								undoPro.dPt[idx] = {spPr: this.utils.reverseSpPr(dpt[idx].spPr,old_dp.spPr)};
						}
					}
				}
			}
		}
		return undoPro;
	}
	
});

File: docsapp/com.ibm.concord.document.common/src/com/ibm/concord/document/common/chart/DataSeries.java
/* ***************************************************************** */
/*                                                                   */
/* IBM Confidential                                                  */
/*                                                                   */
/* IBM Docs Source Materials                                         */
/*                                                                   */
/* (c) Copyright IBM Corporation 2012. All Rights Reserved.          */
/*                                                                   */
/* ***************************************************************** */

package com.ibm.concord.document.common.chart;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import com.ibm.json.java.JSONObject;

public class DataSeries 
{
	private Plot parent;
	private JSONObject content;
	private String id;
	private Map<String, DataSequence> data;
	private int changes;
	
	ChartDocument chart;

	public DataSeries(Plot plot, ChartDocument chart)
	{
		this.parent = plot;
		this.chart = chart;
		data = new HashMap<String, DataSequence>();
	}
	
	public String getId() 
	{
		return id;
	}

	public void setId(String id) 
	{
		this.id = id;
	}
		
	public DataSequence getDataSequence(String role)
	{
		DataSequence dataSeq = data.get(role);
		return dataSeq;
	}
	
	public JSONObject toJson()
	{
		Iterator<Map.Entry<String, DataSequence>> itor = data.entrySet().iterator();
		while(itor.hasNext())
		{
			Map.Entry<String, DataSequence> entry = itor.next();
			String key = entry.getKey();
			DataSequence seq = entry.getValue();
			JSONObject vJson = (JSONObject)content.get(key);
			String addr = seq.getAddress();
			String bacAddr = seq.getBacAddress();
			
			if(addr != null && addr.length() > 0)
			{
				if(vJson == null){
					vJson = new JSONObject();
					content.put(key, vJson);
				}
				//chart.pivotSource!=0 means pivot chart, and addr.charAt(0)=='[' means external file reference
				//Don't remove the cache data for these two cases
				if(chart.pivotSource==0 && addr.charAt(0)!='[')
					vJson.remove(ChartConstant.CACHE);
				vJson.put(ChartConstant.REF, addr);
				if(bacAddr != null)
					vJson.put(ChartConstant.BACREF, bacAddr);
				else
					vJson.remove(ChartConstant.BACREF);
			}
			else if(vJson != null){
				vJson.remove(ChartConstant.REF);
				vJson.remove(ChartConstant.BACREF);
			}
		}
		
		content.put(ChartConstant.CHANGES, changes);
		return content;
	}
	
	public void loadFromJson(JSONObject content)
	{
		this.content = content;
		
		Number cha = (Number)content.get(ChartConstant.CHANGES);
	    if(cha != null)
		   changes = cha.intValue();
		
		DataProvider dataProvider = this.chart.getDataProvider();
		this.id = (String)content.get("id");
		for(int i=0;i<ChartDocument.roles.size();i++)
		{
			String role = ChartDocument.roles.get(i);
			JSONObject json = (JSONObject)content.get(role);
			if(json!=null && dataProvider!=null)
			{
				String ref = (String)json.get(ChartConstant.BACREF);
				if(ref==null || ref.length()==0)
					ref = (String)json.get(ChartConstant.REF);
				
                //Don't create sequence for external file reference
				if(ref != null && ref.length() > 0 && ref.charAt(0)!='[')
				{
					DataSequence seq = dataProvider.createDataSequence(ref);
					data.put(role,seq);
				}
			}
		}
	}
	
	public void set(JSONObject settings)
	{
		JSONObject data = (JSONObject)settings.get("data");
		if(data!=null)
		{
			DataProvider dataProvider = this.chart.getDataProvider();
			Iterator<Map.Entry<String, Object>> itor = data.entrySet().iterator();
			while(itor.hasNext())
			{
				Map.Entry<String, Object> entry = itor.next();
				String role = entry.getKey();
				if(ChartConstant.LABEL.equals(role))
					this.changes |= 0x02;
				else
					this.changes |= 0x01;
				Object pro = entry.getValue();
				if(pro==null || pro instanceof String)
				{
					this.data.remove(role);
					content.remove(role);
				}
				else
				{
					JSONObject json = (JSONObject)pro;
					String ref = (String)json.get(ChartConstant.REF);
					if(ref!=null && ref.length()>0)
					{
						if(dataProvider!=null)
						{
							DataSequence seq = dataProvider.createDataSequence(ref);
							this.data.put(role, seq);
						}
					}
					else
					{
						this.data.remove(role);
						JSONObject roleJson = (JSONObject)content.get(role);
						if(roleJson!=null){
							roleJson.remove(ChartConstant.REF);
							roleJson.remove(ChartConstant.BACREF);
						}						
						else
						{
							roleJson = new JSONObject();
							content.put(role, roleJson);
						}
						if(ChartConstant.LABEL.equals(role))
						{
							String v = (String)json.get("v");
							if(v!=null)
								roleJson.put("v", v);
						}
						else
						{
							JSONObject cache = (JSONObject)json.get(ChartConstant.CACHE);
							if(cache!=null)
								roleJson.put(ChartConstant.CACHE, cache);
						}
					}
				}
			}
		}
		if(settings.containsKey(ChartConstant.MARKER))
		{
			Number marker = (Number)settings.get(ChartConstant.MARKER);
			if(marker == null)
				this.content.remove(ChartConstant.MARKER);
			else
				this.content.put(ChartConstant.MARKER, marker);
		}
		
		if(settings.containsKey(ChartConstant.SMOOTH))
		{
			Number smooth = (Number)settings.get(ChartConstant.SMOOTH);
			if(smooth == null)
				this.content.remove(ChartConstant.SMOOTH);
			else
				this.content.put(ChartConstant.SMOOTH, smooth);
		}
		
		if(settings.containsKey(ChartConstant.DPT))
		{
			changes = changes | 4;
			JSONObject dptSet = (JSONObject)settings.get(ChartConstant.DPT);
			if(dptSet == null)
				this.content.remove(ChartConstant.DPT);
			else
			{
				JSONObject dpt = (JSONObject)this.content.get(ChartConstant.DPT);
				if(dpt == null)
					this.content.put(ChartConstant.DPT, dptSet);
				else
				{
					Iterator<Map.Entry<String, JSONObject>> itor = dptSet.entrySet().iterator();
					while(itor.hasNext())
					{
						Map.Entry<String, JSONObject> entry = itor.next();
						String key = entry.getKey();
						JSONObject pro = entry.getValue();
						
						JSONObject dp =(JSONObject) dpt.get(key);
						if(dp == null)
							dpt.put(key, pro);
						else
						{
							JSONObject spPr = (JSONObject)pro.get(ChartConstant.SPPR);
							if(spPr == null)
								dp.remove(ChartConstant.SPPR);//risk?
							else
							{
								if(dp.containsKey(ChartConstant.SPPR))
									ChartUtils.mergeSpPr((JSONObject)dp.get(ChartConstant.SPPR), spPr);
								else
									dp.put(ChartConstant.SPPR, spPr);
							}
						}
					}
				}
			}
		}
		
		if(settings.containsKey(ChartConstant.SPPR))
		{
			changes = changes | 4;
			JSONObject spPr = (JSONObject)settings.get(ChartConstant.SPPR);
			JSONObject dest = (JSONObject)this.content.get(ChartConstant.SPPR);
			if(spPr == null && dest != null)
			{
				this.content.remove(ChartConstant.SPPR);
			}
			else 
			{
				if(dest == null)
				{
					dest = new JSONObject();
					this.content.put(ChartConstant.SPPR, dest);
				}
				ChartUtils.mergeSpPr(dest, spPr);//merge and maintain change flag
			}
		}
	}
}
